### DISCUSSION TITLE
Which IMMEDIATE_TOKEN rules produce nodes in the tree?

### DISCUSSION THREAD

**Discussion by berchn** [tree-sitter/tree-sitter#1034] [Q&A] [answered]
There are IMMEDIATE_TOKEN rules which produce nodes in the tree but others that don't. The following is an example for C:

```lisp
"#define a(b, c) b+c"
(:TRANSLATION-UNIT ((0 0) (19 0))
 ((:PREPROC-FUNCTION-DEF ((0 0) (19 0))
   ((:|#DEFINE| ((0 0) (7 0)) NIL) ((:NAME :IDENTIFIER) ((8 0) (9 0)) NIL)
    ((:PARAMETERS :PREPROC-PARAMS) ((9 0) (15 0))
     ((:|(| ((9 0) (10 0)) NIL) (:IDENTIFIER ((10 0) (11 0)) NIL)
      (:|,| ((11 0) (12 0)) NIL) (:IDENTIFIER ((13 0) (14 0)) NIL)
      (:|)| ((14 0) (15 0)) NIL)))
    ((:VALUE :PREPROC-ARG) ((15 0) (19 0)) NIL)
    (:|
|
     ((19 0) (19 0)) NIL)))))
```

The left parenthesis is specified as an IMMEDIATE_TOKEN in `src/grammar.json` and produces an unnamed node in the tree. This also appears to be the case if it's switched to TOKEN. In what cases do the token rules produce nodes in the tree?

**Comment by ahlinc** [comment, marked as answer]
All defined tokens should produce at least unnamed nodes in the tree. If they defined as non hidden rules, than they will be represented as named nodes.

Could you provide an example of input where do you have missing nodes?
